### ISSUE TITLE
Media Preview Conflict with Elementor Pro Theme Builder

### ISSUE THREAD

**Issue by reedhewitt** [buddyboss/buddyboss-platform#2797]
**Describe the bug**
When Elementor Pro Theme Builder is in use, photos posted in the timeline are broken for sites that use rewrites instead of symlinks for media.  The problem is that Elementor Pro hooks into the "template_include" filter later than BuddyBoss and it doesn't realize that the media preview URL is anything other than a regular page.  Elementor Pro changes the template file to its own template, which breaks the image preview in BuddyBoss.

Here are the details of the filter added by Elementor Pro:

Filter: template_include
Class: ElementorPro\Modules\ThemeBuilder\Classes\Locations_Manager
Method: template_include
Priority: 11

There are two easy solutions:
1. the places in BuddyBoss which add a filter on the template_include hook can be set to a priority greater than 11, or...
2. the Elementor Pro hook can be removed when the request is identified as a BuddyBoss media preview.

I chose the second option and placed this code in a separate plugin (I removed an additional Elementor hook which was not causing any problems, but I know it is irrelevant for media previews):

```php
add_action('bb_setup_template_for_document_player',  'unhook_elementor_template_include');
add_action('bb_setup_template_for_media_preview', 'unhook_elementor_template_include');
add_action('bb_setup_template_for_video_preview', 'unhook_elementor_template_include');
add_action('bb_setup_template_for_video_thumb_preview', 'unhook_elementor_template_include');

function unhook_elementor_template_include(){
  remove_class_hook('template_include', 'Elementor\Modules\PageTemplates\Module', 'template_include', 11);
  remove_class_hook('template_include', 'ElementorPro\Modules\ThemeBuilder\Classes\Locations_Manager', 'template_include', 11);
}
```

The function remove_class_hook is from https://gist.github.com/tripflex/c6518efc1753cf2392559866b4bd1a53.

After removing that Elementor Pro hook, the images appeared in the timeline and everything looks good.

**To Reproduce**
1. Use the Theme Builder in Elementor Pro to set up default page templates.
2. Post some photos in your timeline via BuddyBoss.
3. Preview images are broken.

**Expected behavior**
The images should appear normally in the timeline regardless of whether Elementor Pro is in use.

**Screenshots**
<img width="607" alt="Screen Shot 2021-08-19 at 2 01 57 AM" src="https://user-images.githubusercontent.com/957141/130023101-8038cae5-9554-4ddc-93d5-3c0e08f59cfd.png">

**Environment**
A development site which is not public accessible at the moment.

**Comment by chetansatasiya** [comment]
@reedhewitt Can you check with this PR https://github.com/buddyboss/buddyboss-platform/pull/2800 it will fix the issues or not?

**Comment by reedhewitt** [comment]
@chetansatasiya Yes indeed, that ought to do it.  Setting the filter priority to PHP_INT_MAX will definitely supersede the paltry 11 that Elementor Pro is using! 😄

The three files modified by PR #2800 are the only three places I know of that would need to be updated.

**Comment by chetansatasiya** [comment]
@reedhewitt Thank you for looking and confirming the issue is fixed and yes, that's 3 places only we need to update the file.

**Comment by reedhewitt** [comment]
@Woolgoose I just tested photos in a forum post and they are working after the changes describe above.  For what it's worth I am not using symlinks for media.

**Comment by Woolgoose** [comment]
@reedhewitt Good to hear that it works after your changes. As I am very new to Github I have a probably quite stupid question. What happens next? Will the next update already include the fix? 
In my case it doesn't make a difference if I use symlink or not. Unchecking it in the BuddyBoss settings does not change the behaviour of showing broken image links.

**Comment by reedhewitt** [comment]
@Woolgoose I'm not one of the BuddyBoss Platform developers, so I don't know for sure when this will get incorporated.  (I'm just a user who reported the issue and the fix that worked for me.)  But looking at the page for pull request #2800, it looks like @chetansatasiya has these changes targeted for release 1.7.7.  I'm not sure what the timeline for that will be.  I'm on 1.7.5 right now, though, so it doubt it will be long.

If you want to try to fix it right away, you can copy my code and put it in your theme's functions.php file.  This way it doesn't require any modifications to the BuddyBoss code and it will not conflict with the fix that they have in the pipeline.  Here is the code that worked for me, including the kind of long `remove_class_hook()` function that is necessary to remove a hook that was added in a class instance which we don't have access to:

```php
add_action('bb_setup_template_for_document_player',  'unhook_elementor_template_include');
add_action('bb_setup_template_for_media_preview', 'unhook_elementor_template_include');
add_action('bb_setup_template_for_video_preview', 'unhook_elementor_template_include');
add_action('bb_setup_template_for_video_thumb_preview', 'unhook_elementor_template_include');

function unhook_elementor_template_include(){
  remove_class_hook('template_include', 'Elementor\Modules\PageTemplates\Module', 'template_include', 11);
  remove_class_hook('template_include', 'ElementorPro\Modules\ThemeBuilder\Classes\Locations_Manager', 'template_include', 11);
}

// Thanks to https://gist.github.com/tripflex/c6518efc1753cf2392559866b4bd1a53
if( ! function_exists( 'remove_class_hook' ) ){
    function remove_class_hook( $tag, $class_name = '', $method_name = '', $priority = 10 ) {
        global $wp_filter;
        if ( ! isset( $wp_filter[ $tag ] ) ) {
            return FALSE;
        }
        if ( is_object( $wp_filter[ $tag ] ) && isset( $wp_filter[ $tag ]->callbacks ) ) {
            $fob       = $wp_filter[ $tag ];
            $callbacks = &$wp_filter[ $tag ]->callbacks;
        } else {
            $callbacks = &$wp_filter[ $tag ];
        }
        if ( ! isset( $callbacks[ $priority ] ) || empty( $callbacks[ $priority ] ) ) {
            return FALSE;
        }
        foreach ( (array) $callbacks[ $priority ] as $filter_id => $filter ) {
            if ( ! isset( $filter['function'] ) || ! is_array( $filter['function'] ) ) {
                continue;
            }
            if ( ! is_object( $filter['function'][0] ) ) {
                continue;
            }
            if ( $filter['function'][1] !== $method_name ) {
                continue;
            }
            if ( get_class( $filter['function'][0] ) === $class_name ) {
                if ( isset( $fob ) ) {
                    $fob->remove_filter( $tag, $filter['function'], $priority );
                } else {
                    unset( $callbacks[ $priority ][ $filter_id ] );
                    if ( empty( $callbacks[ $priority ] ) ) {
                        unset( $callbacks[ $priority ] );
                    }
                    if ( empty( $callbacks ) ) {
                        $callbacks = array();
                    }
                    unset( $GLOBALS['merged_filters'][ $tag ] );
                }
                return TRUE;
            }
        }
        return FALSE;
    }
}
```

**Comment by Woolgoose** [comment]
@reedhewitt Thanks a lot. I will try and report if it works in my case.

**Comment by Woolgoose** [comment]
Works perfectly. Tested it on stage site with all other plugins except Elementor+ PRO and BB platform deactivated and also works on production site with all other plugins activated. Great work. I will suggest that fix also on my original post.

**Comment by reedhewitt** [comment]
@Woolgoose  Great!  I'm glad it works for you.

**Comment by marienel21** [comment]
Hi @reedhewitt @Woolgoose 

The issue mentioned above is released in BuddyBoss Platform v1.7.7

You can check release notes here:
https://www.buddyboss.com/resources/buddyboss-platform-releases/1-7-7/

Please update the plugin to see the latest changes.

Thank you.

**Comment by reedhewitt** [comment]
Thanks, @marienel21!

**Comment by Woolgoose** [comment]
Great! That went fast.

**Comment by anywheresa** [comment]
Still not working for me.
